src/main/java/io/bf2/kafka/authorizer/CustomAclAuthorizer.java

```diff
@@ -246,6 +247,35 @@ void addAllowedListeners(Map<String, ?> configs) {
         }
     }
 
+    /**
+     * To avoid misleading users we delete principals that are explicitly targeted by CustomAclAuthorizer rules.
+     * It is possible for a user with ACL rules in Zookeeper to become a user managed by CustomAclAuthorizer, in which
+     * case we would never use or update rules for them in ZK. The old rulesremain, dangling, in ZK and visible through
+     * kafka ACL tooling. This would mislead users becuase there will be a mismatch between what the ZK rules say they
+     * should be permitted to do and the truth.
+     */
+    private void deleteOrphanBindingsFromDelegate() {
+        List<AclBindingFilter> toDelete = aclPrincipals.stream().map(aclPrincipal -> {
+            AccessControlEntryFilter principalFilter = new AccessControlEntryFilter(aclPrincipal, null, AclOperation.ANY, AclPermissionType.ANY);
+            return new AclBindingFilter(ResourcePatternFilter.ANY, principalFilter);
+        }).collect(Collectors.toList());
+        if (!toDelete.isEmpty()) {
```

Comment:
There is a startup race condition here as each broker in the cluster will attempt to delete orphaned rules. 

In practice that should be a very rare race as we won't roll multiple brokers at the same time but we should probably handle the delete failing because the target AclBindingFilter doesn't exist in the controller.

Reply:
I reproduced this case locally and it seems if the ACL is cached by both brokers both deletion ops to zookeeper result in a successful deletion result on both brokers:

[roby@roby scripts]$ docker logs kas-broker-plugin-kafka | grep orphan -i | grep abd-27
[2022-10-03 22:25:16,827] INFO Orphan deletion: Deletion succeeded for binding (pattern=ResourcePattern(resourceType=TOPIC, name=abd-27, patternType=LITERAL), entry=User:other2 has ALLOW permission for operations: DESCRIBE from hosts: *) (io.bf2.kafka.authorizer.CustomAclAuthorizer)
[roby@roby scripts]$ docker logs kas-broker-plugin-kafka-2 | grep orphan -i | grep abd-27
[2022-10-03 22:25:16,357] INFO Orphan deletion: Deletion succeeded for binding (pattern=ResourcePattern(resourceType=TOPIC, name=abd-27, patternType=LITERAL), entry=User:other2 has ALLOW permission for operations: DESCRIBE from hosts: *) (io.bf2.kafka.authorizer.CustomAclAuthorizer)

Looks like the [conditional delete](https://github.com/apache/kafka/blob/50029d3ed8ba576fd605df6a67f4229303cc944d/core/src/main/scala/kafka/zk/KafkaZkClient.scala#L1341) called on the kafka ZK client succeeds if the znode has already been deleted.